go/libraries/doltcore/sqle/setalgebra/union.go

```diff
@@ -0,0 +1,342 @@
+// Copyright 2020 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package setalgebra
+
+import (
+	"github.com/liquidata-inc/dolt/go/store/hash"
+	"github.com/liquidata-inc/dolt/go/store/types"
+)
+
+// finiteSetUnion adds all points from both sets to a new FiniteSet
+func finiteSetUnion(fs1, fs2 FiniteSet) (FiniteSet, error) {
+	hashToVal := make(map[hash.Hash]types.Value, len(fs1.HashToVal)+len(fs2.HashToVal))
+	for h, v := range fs1.HashToVal {
+		hashToVal[h] = v
+	}
+
+	for h, v := range fs2.HashToVal {
+		hashToVal[h] = v
+	}
+
+	return FiniteSet{HashToVal: hashToVal}, nil
+}
+
+// copyIntervalEndpoint makes a copy of an interval endpoint
+func copyIntervalEndpoint(ep *IntervalEndpoint) *IntervalEndpoint {
+	if ep == nil {
+		return nil
+	}
+
+	copyOf := *ep
+
+	return &copyOf
+}
+
+// finiteSetIntervalUnion will check all the points of the FiniteSet and see which ones are not in the given interval.
+// if all points are in the interval then the resulting set is the interval itself, otherwise a CompositeSet containing
+// the missing points as a new FiniteSet and the interval is returned.
+func finiteSetIntervalUnion(fs FiniteSet, in Interval) (Set, error) {
+	inStart := copyIntervalEndpoint(in.Start)
+	inEnd := copyIntervalEndpoint(in.End)
+
+	hashToVal := make(map[hash.Hash]types.Value, len(fs.HashToVal))
+	for h, v := range fs.HashToVal {
+		inRange, err := ValueInInterval(in, v)
+
+		if err != nil {
+			return nil, err
+		}
+
+		if !inRange {
+			if inStart != nil && !inStart.Inclusive {
+				if inStart.Val.Equals(v) {
+					inStart.Inclusive = true
+					continue
+				}
+			}
+
+			if in.End != nil && !in.End.Inclusive {
+				if in.End.Val.Equals(v) {
+					inEnd.Inclusive = true
+					continue
+				}
+			}
+
+			hashToVal[h] = v
+		}
+	}
+
+	resInterval := Interval{in.nbf, inStart, inEnd}
+	if len(hashToVal) > 0 {
+		newSet := FiniteSet{HashToVal: hashToVal}
+		return CompositeSet{newSet, []Interval{resInterval}}, nil
+	} else {
+		return resInterval, nil
+	}
+}
+
+// finiteSetCompositeSetUnion checks all the points in a FiniteSet against the CompositeSet to find all points not
+// represented in the CompositeSet (So not in any of it's intervals and not in it's existing FiniteSet), and adds those
+// points to the compositeSet
+func finiteSetCompositeSetUnion(fs FiniteSet, cs CompositeSet) (Set, error) {
```

Comment:
Shouldn't this just be Union(cs.Set, Union(fs, cs.Interval))?

Reply:
that could result in values in the finiteset that are already covered by the interval.  In the code that actually uses this that would result in a row being in the result set 2x.